### PR TITLE
[4.1] Allowing adding a variable for the field Label and Description elements

### PR DESCRIPTION
**Pull request by infograf768**
While testing https://github.com/joomla/joomla-cms/pull/20491 I found out that it would be very convenient to allow labels and descriptions strings to use a variable (%s) in order to simplify the multiple use of a value with a `Text::sprintf`.

### Summary of Changes
Added 2 new elements to FormField:
`varLabel` and `varDescription`
Using sprintf when these elements have a value.

Also modified file to use `Text::` instead of `\JText::`

### Testing Instructions
Patch.
Modify an xml. For this example I just modified `/administrator/components/com_content/forms/article.xml`
```
		<field
			name="image_intro"
			type="media"
			label="COM_CONTENT_FIELD_INTRO_LABEL"
			description="COM_CONTENT_FIELD_INTRO_DESCRIPTION"
			varLabel="JYES"
			varDescription="JNO"
		/>
```
Then I modified and added a description string in `/administrator/language/en-GB/en-GB.com_content.ini` line 53-54
```
COM_CONTENT_FIELD_INTRO_DESCRIPTION="Select an image %s"
COM_CONTENT_FIELD_INTRO_LABEL="Intro Image %s"
```
Remark that I added the placeholder for the variable `%s`

### Result
screen shot 2018-05-22 at 08 35 18

### Documentation Changes Required
YES.

Welcome any corrections/suggestions for the comments.

At this stage, this PR allows only one variable. 
It would be great if it was improved to allow many variables so that we can use `%1$s` `%2$s` etc.

### CODE REVIEW

**Comment by SharkyKZ** [comment]
Please use camelCase where applicable.

File: libraries/src/Form/FormField.php [hunk 1]
```diff
@@ -986,11 +1007,35 @@ protected function getLayoutData()
 	{
 		// Label preprocess
 		$label = $this->element['label'] ? (string) $this->element['label'] : (string) $this->element['name'];
-		$label = $this->translateLabel ? \JText::_($label) : $label;
+
+		// Do we have a text variable to use when displaying the label?
+		$var_label = !empty($this->var_label) ? $this->var_label : null;
```
Comment (SharkyKZ) [inline libraries/src/Form/FormField.php]: Suggest to simplify the code:

```
if ($this->varLabel !== '')
{
	$label = $this->translateLabel ? Text::sprintf($label, Text::_($this->varLabel)) : $label;
}
else
{
	$label = $this->translateLabel ? Text::_($label) : $label;
}
```
Same for description.

**Comment by infograf768** [comment]
You mean varDescription and varLabel  all over?

**Comment by SharkyKZ** [comment]
Yes.

**Comment by infograf768** [comment]
camelCase done.

File: libraries/src/Form/FormField.php (hunk 1 above)
Comment (infograf768) [reply libraries/src/Form/FormField.php]: I could simplify, but this use a similar code already and I generally prefer separating stuff for easier reading.
I will if it is judged essential to get this merged.

**Comment by carlitorweb** [comment]
I have tested this item :white_check_mark: successfully on b0f943ae1dc3d2b0459d14646d8877bbe693fd69<br /><br />As you said, if this is accepted, we need extend to more variables <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20525">issues.joomla.org/tracker/joomla-cms/20525</a>.</sub>

File: libraries/src/Form/FormField.php [hunk 2]
```diff
@@ -986,11 +1007,35 @@ protected function getLayoutData()
 	{
 		// Label preprocess
 		$label = $this->element['label'] ? (string) $this->element['label'] : (string) $this->element['name'];
-		$label = $this->translateLabel ? \JText::_($label) : $label;
+
+		// Do we have a text variable to use when displaying the label?
+		$varLabel = !empty($this->varLabel) ? $this->varLabel : null;
+
+		if ($varLabel != null)
```
[the remark this answers is not in the document]
Comment (SharkyKZ) [reply libraries/src/Form/FormField.php]: My previous comment still applies. No need for variable assignments here.

**Comment by Quy** [comment]
Since most descriptions have been removed from labels, let's only check `$varDescription` when there is a description. `$varDescription` can be `0` so check for empty string. Furthermore I would say there is no need for `$varDescription` and `$varLabel`. Here is my proposed code. The same can be done with `$varLabel`.

```
		// Description preprocess
		$description = !empty($this->description) ? $this->description : null;

                if (!empty($description) && $this->translateDescription)
                {
                        // Do we have a text variable to use when displaying the description?
                        if ($this->varDescription !== '')
                        {
                                $description = Text::sprintf($description, $this->varDescription);
                        }
                        else
                        {
                                $description = Text::_($description);
                        }
                }
```

**Comment by infograf768** [comment]
Since Formfield was modified, I have to redo this patch anyway. Will also modify its code.

**Comment by infograf768** [comment]
@Quy @SharkyKZ @carlitorweb 
You now can test with the new code.

**Comment by laoneo** [comment]
For me it is hard to find a use case for this pr, where you have two hard coded values in the XML definition to put into the label or description. Why not directly add them.

**Comment by brianteeman** [comment]
@laoneo the specific use case would be for a plugin like system http headers where we have a large number of terms that should never be translated and it is currently hard to notify tt to ignore those terms

**Comment by Quy** [comment]
`$varDescription` and `$varLabel` are obsolete.

**Comment by infograf768** [comment]
> $varDescription and $varLabel are obsolete.

Hmm, why do we have $hint or $type then? (just for my knowledge).

**Review by Bakual — APPROVED** [review]

File: libraries/src/Form/FormField.php [hunk 3]
```diff
@@ -988,11 +1008,35 @@ protected function getLayoutData()
...
+			{
+				$label = Text::_($label);
+			}
+		}
 
 		// Description preprocess
 		$description = !empty($this->description) ? $this->description : null;
-		$description = !empty($description) && $this->translateDescription ? Text::_($description) : $description;
+
+		if (!empty($description) && $this->translateDescription)
```
Comment (Quy) [inline libraries/src/Form/FormField.php:1247]: Since we already tested `!empty`, then we don't need to test `!empty` again...right? 

Change it to: `if ($description && $this->translateDescription)`

I am not sure about `$varDescription` and `$varLabel` so keep them as is?

Comment (infograf768) [reply libraries/src/Form/FormField.php:1247]: > Since we already tested !empty, then we don't need to test !empty again...right?

I do not think so. $description may be null.

> I am not sure about $varDescription and $varLabel so keep them as is?

Yes we have to.

Comment (Quy) [reply libraries/src/Form/FormField.php:1247]: > $description may be null.

Then the if statement would not be executed.

Comment (infograf768) [reply libraries/src/Form/FormField.php:1247]: I don't understand. Indeed when $description is null, the if statement is not executed which is what we want.

Comment (Quy) [reply libraries/src/Form/FormField.php:1247]: `if (!empty($description) && $this->translateDescription)`
should be the same as:
`if ($description && $this->translateDescription)`

It is only a suggestion to simplify the code.

Comment (Bakual) [reply libraries/src/Form/FormField.php:1247]: `if (!empty($description))` and `if ($description)` do the same thing. The difference is only that `!empty()` doesn't give a notice when the property/variable doesn't exist.
Technically, we wouldn't need `!empty()` anywhere here in the class, since the properties and variables always exist (they're properly declared).
In the end it is mainly personal preference which is used. Some prefer to always use `!empty()` and others prefer using the simple boolean check.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 26059d83178490ad0a54cd968a9c7acf09d21ee6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20525">issues.joomla.org/tracker/joomla-cms/20525</a>.</sub>

**Comment by jreys** [comment]
I have tested this item :white_check_mark: successfully on 26059d83178490ad0a54cd968a9c7acf09d21ee6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20525">issues.joomla.org/tracker/joomla-cms/20525</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20525">issues.joomla.org/tracker/joomla-cms/20525</a>.</sub>

**Comment by infograf768** [comment]
Drone restarted.

**Comment by infograf768** [comment]
Restarted drone again... Failure is not due to this PR.
@webdongle @laoneo Can you merge this?

**Comment by Webdongle** [comment]
@infograf768 I do not (and never have) had ability to merge on github.  Also I have resigned from bugsquad.

**Comment by infograf768** [comment]
oops, sorry. I meant @wilsonge 😃

File: libraries/src/Form/FormField.php [hunk 4]
```diff
@@ -988,11 +1008,35 @@ protected function getLayoutData()
 	{
 		// Label preprocess
 		$label = $this->element['label'] ? (string) $this->element['label'] : (string) $this->element['name'];
-		$label = $this->translateLabel ? Text::_($label) : $label;
+
+		if ($this->translateLabel)
+		{
+			// Do we have a text variable to use when displaying the label?
+			if ($this->varLabel !== '')
```
Comment (roland-d) [inline libraries/src/Form/FormField.php:1234]: We can remove the else statement here by putting the `$label = Text::_($label);` before the if statement or if it is more likely that `$this->varLabel` has a value reverse the condition and put the `$label = Text::sprintf($label, Text::_($this->varLabel));` before the if and the `$label = Text::_($label);` inside the if.

So either one of these:
```
$label = Text::_($label);

if ($this->varLabel !== '')
{
				$label = Text::sprintf($label, Text::_($this->varLabel));
			}
```

```
$label = Text::sprintf($label, Text::_($this->varLabel));

if ($this->varLabel === '')
{
			$label = Text::_($label);	
			}
```

File: libraries/src/Form/FormField.php [hunk 5]
```diff
@@ -988,11 +1008,35 @@ protected function getLayoutData()
...
+		}
 
 		// Description preprocess
 		$description = !empty($this->description) ? $this->description : null;
-		$description = !empty($description) && $this->translateDescription ? Text::_($description) : $description;
+
+		if (!empty($description) && $this->translateDescription)
+		{
+			// Do we have a text variable to use when displaying the description?
+			if ($this->varDescription !== '')
```
Comment (roland-d) [inline libraries/src/Form/FormField.php:1250]: Here the same as above, we can remove the else statement.

**Comment by brianteeman** [comment]
it has been a year - will this be merged or not? Is it waiting on something

**Comment by wilsonge** [comment]
Not as it is. I don't want to have this forced single string. The concept is almost there but this needs to be able to take a list of items otherwise we're just going to hit massive limitations. Trying to sort this in a vaguely b/c way for form fields is super complicated tho :(

**Comment by roland-d** [comment]
Just my 2cents, if you need variables in your language strings, you may as well set a new description in the code using something like `$this->form->setFieldAttribute('description, 'a new string')`.

Other than that I would say we can close this as it can be solved in a different and more flexible way.

**Comment by flo-the-cat** [comment]
I have tested this item :red_circle: unsuccessfully on 6c96029af5b94b5d5f1bd5a5546d5f1a6d0e36ec<br /><br />This needs to be looked at again according to the comments made. Would be worth revisiting and taking in to account @George and @Rolands comments<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20525">issues.joomla.org/tracker/joomla-cms/20525</a>.</sub>

**Comment by particthistle** [comment]
@flo-the-cat tagged the wrong George. @wilsonge is this one still on your radar?

@infograf768 your thoughts on whether it's still needed in light of the conversation in 2019?

Or per @roland-d comment should it be closed.

_Commenting on this PR as part of a "Outstanding PR Status List" exercise to work on reducing the number of outstanding PR that have been otherwise orphaned, neglected or forgotten instead of being merged. See Glip Bugs & Fun @ Home channel for more information._

**Comment by infograf768** [comment]
I still think this would be a nice improvement to avoid wrong translation of some strings that should not be translated.

Read https://github.com/joomla/joomla-cms/pull/20525#issuecomment-391345240

I don't think @roland-d proposal solves the issue and fail to understand @wilsonge comment.

**Comment by brianteeman** [comment]
@infograf768 I think the issue is that the example you gave  in the first post doesnt really show a good example of how this would be used. If you post an example using one of the com_csp forms I think people might understand the need for this more.

**Comment by infograf768** [comment]
Example with label (in these cases, sprintf is not used.)

Modifying xml to add `varLabel` in the 2 following cases in `/administrator/components/com_csp/config.xml`

```
		<field
			name="contentsecuritypolicy"
			type="radio"
			label="COM_CSP_CONTENTSECURITYPOLICY"
			layout="joomla.form.field.radio.switcher"
			default="0"
			varLabel="href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank' rel='noopener noreferrer'>Content Security Policy (CSP)"
			>
			<option value="0">JDISABLED</option>
			<option value="1">JENABLED</option>
		</field>
[...]
		<field
			name="frame_ancestors_self_enabled"
			type="radio"
			label="COM_CSP_CONTENTSECURITYPOLICY_FRAME_ANCESTORS_SELF_ENABLED"
			description="COM_CSP_CONTENTSECURITYPOLICY_FRAME_ANCESTORS_SELF_ENABLED_DESC"
			layout="joomla.form.field.radio.switcher"
			default="1"
			showon="contentsecuritypolicy:1[AND]contentsecuritypolicy_mode!:detect"
			varLabel="frame-ancestors 'self'"
			>
			<option value="0">JDISABLED</option>
			<option value="1">JENABLED</option>
		</field>
```
Original strings:

`; Please do not translate the following language string`
`COM_CSP_CONTENTSECURITYPOLICY="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank' rel='noopener noreferrer'>Content Security Policy (CSP)</a>"`

```
; Please do not translate the following language string
COM_CSP_CONTENTSECURITYPOLICY_FRAME_ANCESTORS_SELF_ENABLED="frame-ancestors 'self'"
```

Modified strings:
```
COM_CSP_CONTENTSECURITYPOLICY="<a %s </a>"
COM_CSP_CONTENTSECURITYPOLICY_FRAME_ANCESTORS_SELF_ENABLED="%s"
```

Result
Screen Shot 2021-01-15 at 11 08 38

For a string like
```
; Please do not translate 'max-age' in the following language string
PLG_SYSTEM_HTTPHEADERS_HSTS_MAXAGE_DESC="This option sets the time for 'max-age', it is specified in seconds. The default value is 31536000, which corresponds to one year"
```
the sprintf can be used with `varDescription` and we would have

```
PLG_SYSTEM_HTTPHEADERS_HSTS_MAXAGE_DESC="This option sets the time for '%s', it is specified in seconds. The default value is 31536000, which corresponds to one year"
```

If this PR is improved to allow multiple variables, we could also simplify a string like
```
; Please do not translate 'Content-Security-Policy' & 'Content-Security-Policy-Report-Only' in the following language string
COM_CSP_CONTENTSECURITYPOLICY_REPORT_ONLY_DESC="Use the header 'Content-Security-Policy-Report-Only' instead of 'Content-Security-Policy'."
```

**Comment by brianteeman** [comment]
Thank you - I think people should be able to see the benefit of this now

**Comment by rdeutz** [comment]
I am moving this to 4.1, we are in beta and this is a new feature.

**Comment by bembelimen** [comment]
So this did not make it in 4.0 and from my point of view will not make it into 4.1 either, although I see the benefit of the whole idea the limitations stated by @wilsonge are still valid (especially that you can only add one variable...) and TBH for me it feels super complicated when strings are splitted up in several %s, too.

I think this could be reconsidered, when we would go for "custom named placeholders" to make language keys more clear. (See: https://github.com/joomla/joomla-cms/blob/4.1-dev/libraries/src/Language/Text.php#L113-L122)
